scratch/evaluate_scheduler_qd_dense_scenario.cc

```diff
@@ -0,0 +1,825 @@
+
+/** Simulation Arguments **/
+string applicationType = "onoff";             /* Type of the Tx application */
+string socketType = "ns3::UdpSocketFactory";  /* Socket Type (TCP/UDP) */
+string schedulerType = "ns3::CbapOnlyDmgWifiScheduler";   /* The type of scheduler to be used */
+string phyMode = "DMG_MCS12";                 /* The MCS to be used at the Physical Layer. */
+uint32_t packetSize = 1448;                   /* Application payload size [bytes]. */
+string tcpVariant = "NewReno";                /* TCP Variant Type. */
+uint32_t maxPackets = 0;                      /* Maximum Number of Packets */
+uint32_t msduAggregationSize = 7935;          /* The maximum aggregation size for A-MSDU [bytes]. */
+uint32_t mpduAggregationSize = 262143;        /* The maximum aggregation size for A-MPDU [bytes]. */
+double simulationTime = 10;                   /* Simulation time [s]. */
+bool csv = false;                             /* Enable CSV output. */
+uint8_t allocationId = 1;                     /* The allocation ID of the DMG Tspec element to create */
+
+/* NUmber of SP allocations */
+uint16_t numSPs = 6;
+
+/*** Beamforming CBAP ***/
+uint16_t biThreshold = 10;                    /* BI Threshold to trigger TxSS TXOP. */
+map<Mac48Address, uint16_t> biCounter;   /* Number of beacon intervals that have passed. */
+
+/**  Applications **/
+CommunicationPairList communicationPairList;  /* List of communicating devices. */
+
+/* MAC layer Statistics */
+typedef map<Mac48Address, uint64_t> PacketCountMap;
+PacketCountMap macTxDataFailed;
+PacketCountMap macTxDataOk;
+PacketCountMap macRxDataOk;
+
+Ptr<DmgApWifiMac> apWifiMac;
+
+template <typename T>
+string to_string_with_precision (const T a_value, const int n = 6)
+{
+  ostringstream out;
+  out.precision (n);
+  out << fixed << a_value;
+  return out.str ();
+}
+
+vector<string>
+SplitString (const string &str, char delimiter)
+{
+  stringstream ss (str);
+  string token;
+  vector<string> container;
+
+  while (getline (ss, token, delimiter))
+    {
+      container.push_back (token);
+    }
+  return container;
+}
+
+void
+EnableMyTraces (vector<string> &logComponents, Time tLogStart, Time tLogEnd)
+{
+  for (size_t i = 0; i < logComponents.size (); ++i)
+    {
+      const char* component = logComponents.at (i).c_str ();
+      if (strlen (component) > 0)
+        {
+          NS_LOG_UNCOND ("Logging component " << component);
+          Simulator::Schedule (tLogStart, &LogComponentEnable, component, LOG_LEVEL_ALL);
+          Simulator::Schedule (tLogEnd, &LogComponentDisable, component, LOG_LEVEL_ALL);
+        }
+    }
+}
+
+double
+CalculateSingleStreamThroughput (Ptr<PacketSink> sink, uint64_t &lastTotalRx, double &averageThroughput)
+{
+  double thr = (sink->GetTotalRx () - lastTotalRx) * (double) 8 / 1e5;     /* Convert Application RX Packets to Mbits. */
+  lastTotalRx = sink->GetTotalRx ();
+  averageThroughput += thr;
```

Comment:
Sorry Mattia, this comes from Hany's code and I've never checked it. Fortunately, this code just displays the throughput every 100 ms in the standard output. Any issue won't affect the results we collected for the NIST's report. According to my comment above this should thr in Mbps. Let me know :)

Reply:
ok so the throughput over 100 ms windows is being added.. doesn't look like an average..
it is saved into `CommunicationPair.throughput` so I guess this is not what he wants to do.
I'd say either of these would be ok:
1. rename variables in a more meaningful way
2. remove that field from `CommunicationPair` since it is not used (but it should be done for all scratch scripts..)
3. compute and the average throughput
An average can be updated online if you know how many elements you have been averaging.
Starting from i=0, T_{0,avg}=0, you can compute T_{n,avg} = (n-1)/n * T_{n-1,avg} + T_n/n, for each n>1. In this case you could either save and update `i` as a global variable or compute it as `i= Simulator::Now () / MilliSeconds(100)`